tracer/src/instruction/test.rs

```diff
@@ -1,89 +1,246 @@
+        } else {
+            (rng.next_u32() as u64, (rng.next_u32() & 0x0fff) as i64, 0)
+        };
+
+        // Construct instruction
+        // Since we need control over operands we use a custom `from_operands` method
+        let instruction: I = if is_memory {
+            let ty = TypeId::of::<I>();
+            if ty == TypeId::of::<SB>() || ty == TypeId::of::<SH>() {
+                // Store → FormatS downcast
+                unsafe {
+                    let ctor: fn(usize, usize, i64, u64) -> I =
+                        std::mem::transmute(<SB>::from_operands as fn(_, _, _, _) -> SB);
+                    ctor(rs1_idx as usize, rs2_or_rd_idx as usize, imm, addr_tag)
+                }
+            } else {
+                // Load → FormatLoad downcast
+                unsafe {
+                    let ctor: fn(usize, usize, i64, u64) -> I =
+                        std::mem::transmute(<LB>::from_operands as fn(_, _, _, _) -> LB);
+                    ctor(rs2_or_rd_idx as usize, rs1_idx as usize, imm, addr_tag)
+                }
+            }
+        } else {
+            let ty = TypeId::of::<I>();
+            if ty == TypeId::of::<DIV>()
+                || ty == TypeId::of::<DIVU>()
+                || ty == TypeId::of::<REM>()
+                || ty == TypeId::of::<REMU>()
+            {
+                let rd_idx = rng.next_u64() % REG_COUNT as u64;
+                // Div → FormatR downcast
+                unsafe {
+                    let ctor: fn(usize, usize, usize, u64) -> I =
+                        std::mem::transmute(<DIV>::from_operands as fn(_, _, _, _) -> DIV);
+                    ctor(
+                        rd_idx as usize,
+                        rs1_idx as usize,
+                        rs2_or_rd_idx as usize,
+                        addr_tag,
+                    );
+                }
+            }
+
+            let word = rng.next_u32();
+            I::new(word, addr_tag, false) // non-memory cases can use normal constructor
+        };
+
+        // Seed registers
+        cpu_exec.x[rs1_idx as usize] = rs1_val as i64;
+        cpu_trace.x[rs1_idx as usize] = rs1_val as i64;
+
+        if is_memory
+            && (TypeId::of::<I>() == TypeId::of::<SB>() || TypeId::of::<I>() == TypeId::of::<SH>())
+        {
+            let rs2_val = rng.next_u32() as i64;
+            cpu_exec.x[rs2_or_rd_idx as usize] = rs2_val;
+            cpu_trace.x[rs2_or_rd_idx as usize] = rs2_val;
+        }
+
+        // Execute the (random) seeded instructions
+        let mut ram_access = <I as RISCVInstruction>::RAMAccess::default();
+        instruction.execute(&mut cpu_exec, &mut ram_access);
+        instruction.trace(&mut cpu_trace);
+
+        // The first 32 registers should have the same state after execution
+        for i in 0..31 {
+            let real_reg = cpu_exec.x[i] as i32; // we cast due to sign extensions
+            let virtual_reg = cpu_trace.x[i] as i32;
```

Comment:
nit: let's call these `expected_value` and `virtual_sequence_value` or something to avoid confusion with the virtual registers `cpu.x[32..]`